### PR TITLE
Use composer for dependencies

### PR DESCRIPTION
**Pull request by mbirth**
I replaced most dependencies by a proper composer.json incl. Composer's autoloader. I compared the existing ones to get the most perfect matching official version. However, with phpseclib I had to use a more recent version to be compatible with PHP7. But the needed changes were minimal. This also sets the foundation for proper phpunit tests.

I tested installation (using a SQLite database) and retrieve.php and found no problems.

Some dependencies weren't listed in Packagist (Composer's repository), so I added them and they're hosted on my GitHub account. These could be moved over to yours in the future, so you can maintain them.

### CODE REVIEW

**Comment by mesa57** [comment]
Could it be possible to integrate composer in install.php ?
I do not think users will like it to execute 2 seperate steps...

**Comment by mbirth** [comment]
It [seems](http://stackoverflow.com/questions/17219436/run-composer-with-a-php-script-in-browser) [possible](https://github.com/composer/composer/issues/1906). But it's everything but practical. And it's only a matter of time until the current workarounds break. So I wouldn't advise it.

Also, Composer is something everyone interested in PHP should know by now and those 2 additional steps (1 step if you're updating) shouldn't hurt.

**Comment by mesa57** [comment]
The bulk of our users are not interested in PHP let alone they understand what it is.
Spotweb is pure an application package and install should be as simple as possible.
I was in the wrong impression that composer only would manage dependency's during development.

**Comment by mbirth** [comment]
You could offer releases for that, i.e. ready-to-use .zips with all dependencies included.

**Comment by mesa57** [comment]
People are used to clone from gtihub. 
We have a large community of DSM users who do it that way.
Also there is the question if they would be able to install composer.

Can we not use it only at dev time and leave spotweb as it is now on github ?

**Comment by mesa57** [comment]
@Sweepr Please also take a look into this.

**Comment by mbirth** [comment]
Composer works where PHP is installed. So using it on a DSM is no problem - you just have to SSH into your DSM and run it from there. (I just tested it.)

There are some people who use Composer but also commit all external libraries to their git repo. This solves [such problems](http://lornajane.net/posts/2014/using-composer-without-gitignoring) but also creates [new ones](https://getcomposer.org/doc/faqs/should-i-commit-the-dependencies-in-my-vendor-directory.md).

**Comment by Sweepr** [comment]
Ok so far so good, i've got it working under Xampp but this is far from dummy proof and as Mesa57 stated, we should look into making the install easier.

Question: why download the vendor folder, why not put it in the branch by default?

# php -r "copy('https://getcomposer.org/installer', 'composer-setup.php');"

# php composer-setup.php  --install-dir=c:\xampp\htdocs\master
All settings correct for using Composer
Downloading 1.2.0...

Composer successfully installed to: C:\xampp\htdocs\master\composer.phar
Use it: php c:\xampp\htdocs\master/composer.phar

# php composer.phar install --no-dev
Loading composer repositories with package information
Installing dependencies from lock file
  - Installing abraham/twitteroauth (0.2.1)
    Downloading: 100%

  - Installing boxcar/boxcar (dev-master 1357299)
    Cloning 1357299c01bca1da57391f75a5de8214d927bd4d

  - Installing boxcar/boxcar (dev-master 1357299)
    Downloading: 100%

  - Installing digitalstratum/fts_parser (dev-spotweb b693bfb)
    Cloning b693bfb6f348a1875dd8edd512cce360dceb1138
  
  - Installing digitalstratum/fts_parser (dev-spotweb b693bfb)
    Downloading: 100%

  - Installing jean-damien/pchart (v2.1.4)
    Downloading: 100%

  - Installing mobiledetect/mobiledetectlib (2.5.9)
    Downloading: 100%

  - Installing notifo/notifo-php (dev-master 3f5faee)
    Cloning 3f5faee77e6c79a769de2577ff580aaf679543a8

  - Installing notifo/notifo-php (dev-master 3f5faee)
    Downloading: 100%

  - Installing phpseclib/phpseclib (2.0.2)
    Downloading: 100%

  - Installing uskr/nma-php (v0.0.4)
    Downloading: 100%

  - Installing xenji/prowlphp (1.0.3)
    Downloading: 100%

Generating autoload files... 

Normal install went fine, retrieving spots..

Install for Xampp is confirmed.

If there is someone who can test this on a DSM or any other NAS would be great, if not i'll try if i can tomorrow.

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+    "name": "Spotweb",
```
Comment (GeoffreyDijkstra) [inline composer.json]: Drop the `name` keyword this is meant for composer packages to be identifiable and composer should not recognize this as a package.

**Comment by GeoffreyDijkstra** [comment]
@Sweepr @mbirth I already run spotweb with composer from my own fork. And as long as you have PHP running on a NAS it's as easy as `git clone git@github.com:spotweb/spotweb.git` and  `composer.phar install`.

@Sweepr I wouldn't go as far to include the vendor. I will not give a whole lot of reasons because "this is what composer is for" :)

It is easy to make a script which makes zip files of a github project including the vendor folder etc. If you guys like I can create a small example script which checks out the latest release runs composer install in that folder and zips the whole content (I did this before).

@mesa57 I also suggest to start using release numbers on spotweb (git tags).

**Comment by GeoffreyDijkstra** [comment]
@Sweepr @mesa57 @mbirth This is a basic way how you can create a release based on branches (not recommended) and tags (the recommended way): https://github.com/GeoffreyDijkstra/demo-git-releases Using an existing php implementation of the github API and some basic php with linux commands.

**Comment by mbirth** [comment]
Any progress here?

**Comment by GeoffreyDijkstra** [comment]
@mesa57 I made some changes to this PR (by making a new PR to @mbirth), namely that the vendor folder is now included, so the community can just clone Spotweb without thinking about composer. We now need to wait for @mbirth to fix those changes, but for now you can see how this branch looks like here https://github.com/GeoffreyDijkstra/spotweb/tree/feature/composer This way developers can use composer add dev time.

**Comment by mesa57** [comment]
@GeoffreyDijkstra : could you carefully review this pull request, As I said earlier, I do not have experience with composer, but non spotweb dev's should not be forced to install it.

**Comment by GeoffreyDijkstra** [comment]
@mesa57 I agree, for users the installation should be as easy as possible and not to worry about extra tools. I will do a review and run a final test to see if everything is working as expected.

**Comment by GeoffreyDijkstra** [comment]
@mesa57 this PR includes the revert of retriever.php #199, I also walked through the whole install process etc, I found a few bugs and fixed those by updating the packages (that's why we love composer).

**Comment by mbirth** [comment]
Woah, finally! :+1:

**Pull request merged**